apps/docs/pages/guides/auth/auth-postgres-hooks.mdx

```diff
@@ -0,0 +1,851 @@
+declare
+  -- Add your variables here
+begin
+  -- Add your logic here
+  return response
+end;
+$$ language plpgsql;
+```
+
+Note that the JSONB operator `?` can't be used when writing a hook, because it is a reserved character in the ORM used by Supabase Auth.
+
+### Handle Hook errors
+
+Unless otherwise specified, your hook should return errors with the following fields:
+
+- `error` An object that contains information about the error.
+    - `http_code` A number indicating the HTTP code to be returned. If not set, the code is HTTP 500 Internal Server Error.
+    - `message` A message to be returned in the HTTP response. Required.
+
+You should return an error if an input in the `event` payload has been processed incorrectly. This error could result from a server error (for example, an unexpected error with the data) or a user error (for example, the user has exceeded the number of available attempts).
+
+In general, Supabase Auth ignores all fields other than the `error` field attached to the returned JSON.
+
+Example output:
+```jsx
+{
+  "error": {
+    "http_code": 429,
+    "message": "You can only verify a factor once every 10 seconds."
+  }
+}
+```
+
+### Assign Hook permissions
+
+You need to assign additional permissions to the hook and the tables that it interacts with.
+
+By default, Supabase Auth invokes hooks with the `supabase_auth_admin` role. This role doesn't have permissions to modify the `public` schema. You need to grant the role permission to execute your hook function:
+
+```sql
+grant execute
+  on function public.custom_access_token_hook
+  to supabase_auth_admin;
+```
+
+Also revoke permissions from the `authenticated` and `anon` roles to ensure the function is not accessible by Supabase Serverless APIs.
+
+```sql
+revoke execute
+  on function public.custom_access_token_hook
+  from authenticated, anon;
+```
+
+<Admonition type="caution">
+Avoid adding a `security definer` tag to the function to avoid permission errors.
+</Admonition>
+
+At time of writing, Postgres functions have an implicit tag of `security invoker` and run with the permissions of the calling user. In the case of Auth Hooks, `supabase_auth_admin`  is the invoker and hook functions run with the permissions of `supabase_auth_admin`.
+
+The `security definer` tag executes a function with the privileges of the user who owns it.  In the case of functions created via the Supabase Dashboard, this user is `postgres`, which does not have permissions for the Auth schema. Using `security definer` tag would prevent Supabase Auth from having the appropriate permissions to access tables or execute functions.
+
+
+### Timeouts
+
+Ensure that your hooks complete within **2 seconds** to avoid any errors.
+
+
+
+### Connect a Hook
+
+In the dashboard, navigate to [`Authentication > Hooks (Beta)`](/dashboard/project/_/auth/hooks) and select the appropriate PostgreSQL function from the dropdown menu.
+
+
+## MFA Verification Attempt Extensibility Point
```

Comment:
Headings should be sentence case:

```suggestion
## Hook: MFA verification attempt
```

Also in general, I don't think I like the `extensibility point` language a lot. It's long and hard to pronounce, why not just use `hook` everywhere?

Reply:
The main driver towards using the `extensibility point` or `execution point` was to have a way to distinguish between the hook (specific implementation) vs the location at which the hook is called. 

For instance, one might not necessarily use the `Custom Access Token Hook` to mint access tokens even though that is the main use case we anticipate. In this case the `points` terminology provides a way to distinguish between the set of possible cases (`Point`)  vs the specific implementation (`Custom Access Token Hook`)

Open to suggestions if there is a simpler way of doing so though.